src/views/Graph.vue

```diff
@@ -394,6 +509,170 @@ export default {
   },
 
   methods: {
+    createNode (nodeName, cyclePoint) {
+      // const abc = this.workflows[0].tokens.clone({cycle: "20240624T1200Z", workflow: "consolidate_observations"})
+      // adds a new node object to 'nodes' array
+      const nodePath = this.workflowIDs[0]
+      const nodeId = `${nodePath}//${cyclePoint}/${nodeName}`
+      const nodeParent = `${nodePath}//${cyclePoint}`
+      const newNode = {
+        children: [],
+        familyTree: undefined,
+        id: nodeId,
+        name: nodeName,
+        node: {
+          firstParent: {
+            cyclePoint,
+            id: `${nodeParent}/root/${nodeName}`,
+            name: 'root',
+            state: 'succeeded'
+          },
+          id: false,
+          isHeld: false,
+          isQueued: false,
+          isRunahead: false,
+          name: nodeName,
+          state: 'succeeded',
+          task: {
+            meanElapsedTime: 0
+          }
+        },
+        parent: nodeParent,
+        tokens: {
+          cycle: cyclePoint,
+          edge: undefined,
+          id: nodeId,
+          job: undefined,
+          namespce: undefined,
+          relativeID: `${cyclePoint}/${nodeName}`,
+          task: nodeName,
+          user: undefined,
+          workflow: undefined,
+          workflowID: nodePath
+        },
+        type: 'family'
+      }
+      return newNode
+    },
+    removeNode (nodeName, cyclePoint, nodes) {
+      // removes a node object from the 'nodes' array
+      const nodePath = this.workflowIDs[0]
+      const nodeId = `${nodePath}//${cyclePoint}/${nodeName}`
+      const nodesFiltered = nodes.filter((node) => {
+        return node.id !== nodeId
+      })
+      return nodesFiltered
+    },
+    checkForNode (nodeName, cyclePoint, nodes) {
+      // checks the 'edges' array to see if edge is present
+      const nodePath = this.workflowIDs[0]
+      const nodeId = `${nodePath}//${cyclePoint}/${nodeName}`
+      const nodeSearch = nodes.find((node) => { return node.id === nodeId })
+      if (nodeSearch === -1) {
+        return null
+      } else {
+        return nodeSearch
+      }
+    },
+    createEdge (sourceName, targetName, sourceCyclePoint, targetCyclePoint) {
+      // adds a new edge object to 'edges' array
+      const edgePath = this.workflowIDs[0]
+      const edgeId = `${edgePath}//$edge|${sourceCyclePoint}/${sourceName}|${targetCyclePoint}/${targetName}`
```

Comment:
If you can pass this function the source and target tokens rather than their IDs, then this becomes a lot eaisier.

I think we can do it like this:

```js
// untested
const edge_tokens = src_tokens.clone(cycle=`\$edge|${src_tokens.relativeID}|${tgt_tokens.relativeID}`, task=null, job=null)
const edge_tokens.edge = [src_tokens, tgt_tokens]
const edge_id = edge_tokens.id
```